extensions/wikia/Email/MobileApplicationsLinksGenerator.class.php

```diff
@@ -0,0 +1,174 @@
+				$mobileApplicationsLinks =
+					$this->traverseThrough( $mobileApplications, $storeLanguage );
+			}
+
+			$mobileApplicationsLinks =
+				$this->fillWithWikiaAccountLinksIfEmpty( $mobileApplicationsLinks, $storeLanguage );
+
+			// Even if response is not valid (for example there are some difficulties with mobile applications service)
+			// result will be cached to prevent too long delays in email generation
+			$wgMemc->set( $this->createCacheKeyForMobileApplicationsLinks(),
+				$mobileApplicationsLinks, time() + self::MOBILE_APPLICATIONS_LINKS_EVICTION_TIME );
+		}
+
+		return $mobileApplicationsLinks;
+	}
+
+	/**
+	 * Fills array with wikia mobile application store link if application was not found
+	 *
+	 * @param array $mobileApplicationsLinks
+	 * @param string $storeLanguage
+	 * @return array with links to app store and google play
+	 */
+	private function fillWithWikiaAccountLinksIfEmpty( $mobileApplicationsLinks, $storeLanguage ) {
+		if ( empty ( $mobileApplicationsLinks[self::ANDROID_PLATFORM] ) ) {
+			$mobileApplicationsLinks[self::ANDROID_PLATFORM] =
+				sprintf( self::WIKIA_GOOGLE_PALY_URL,
+					$this->prepareLanguageForGooglePlay( $storeLanguage ) );
+		}
+
+		if ( empty( $mobileApplicationsLinks[self::IOS_PLATFORM] ) ) {
+			$mobileApplicationsLinks[self::IOS_PLATFORM] =
+				sprintf( self::WIKIA_APP_STORE_URL,
+					$this->prepareLanguageForAppStore( $storeLanguage ) );
+		}
+
+		return $mobileApplicationsLinks;
+	}
+
+	/**
+	 * @param string $language
+	 * @return string
+	 */
+	private function prepareLanguageForGooglePlay( $language ) {
+		return $language;
+	}
+
+	/**
+	 * @param string $language
+	 * @return string
+	 */
+	private function prepareLanguageForAppStore( $language ) {
+		return empty ( self::$appStoreLanguagesMapping[$language] ) ? 'us'
+			: self::$appStoreLanguagesMapping[$language];
+	}
+
+	private function createCacheKeyForMobileApplicationsLinks() {
+		return wfMemcKey( 'mobileApplicationsLinks' );
+	}
+
+	/**
+	 * @return string or boolean (because of Http::request)
+	 */
+	private function fetchMobileApplicationsDetails() {
+		// currently it does not matter if Android or iOS value is added, data is returned for both Android and iOS
+		return Http::request( 'GET',
+			'https://services.wikia.com/mobile-applications/platform/android' );
+	}
+
+	/**
+	 * @param string response - raw, not parsed response
+	 *
+	 * @return boolean
+	 */
+	private function applicationsExistIn( $response ) {
+		return strpos( $response, 'wikia_id":' . \F::app()->wg->CityId ) !== false;
+	}
+
+	/**
+	 * @param array $mobileApplications - array containing information about mobile applications
+	 * @param string $storeLanguage
+	 * @return array - containing mobile applications links
+	 */
+	private function traverseThrough( $mobileApplications, $storeLanguage ) {
+		$mobileApplicationsLinks = [];
+		$siteId = \F::app()->wg->CityId;
+
+		foreach ( $mobileApplications['apps'] as $app ) {
```

Comment:
I know this is a copy/paste, but the indentation level is pretty deep here.  Consider splitting the contents of each loop into their own method.